modules/javafx.graphics/src/main/native-glass/gtk/glass_dnd.cpp

```diff
@@ -760,132 +734,346 @@ static gboolean dnd_source_set_raw(GtkWidget *widget, GtkSelectionData *sel_data
+    } else if (selected & GDK_ACTION_LINK) {
+        cursor = gdk_cursor_new_from_name(gdk_display_get_default(), "dnd-link");
+    }
+
+    if (cursor == NULL) {
+        cursor = gdk_cursor_new_from_name(gdk_display_get_default(), "dnd-none");
+    }
+
+    return cursor;
+}
+
+static void process_dnd_source_drag_status(DragSourceContext *ctx, GdkEvent *event) {
+    GdkEventDND *eventDnd = &event->dnd;
+    GdkDragAction selected = gdk_drag_context_get_selected_action(eventDnd->context);
+    gdk_pointer_grab(ctx->dnd_window, FALSE, GDK_MOUSE_EVENTS_MASK, NULL,
+                        get_action_cursor(selected), GDK_CURRENT_TIME);
+}
+
+static void add_target_from_jstring(JNIEnv *env, GList **list, jstring string) {
+    const char *gstring = env->GetStringUTFChars(string, NULL);
+    if (g_strcmp0(gstring, "text/plain") == 0) {
+        *list = g_list_append(*list, TARGET_UTF8_STRING_ATOM);
+        *list = g_list_append(*list, TARGET_MIME_TEXT_PLAIN_ATOM);
+        *list = g_list_append(*list, TARGET_STRING_ATOM);
+        *list = g_list_append(*list, TARGET_COMPOUND_TEXT_ATOM);
+    } else if (g_strcmp0(gstring, "application/x-java-rawimage") == 0) {
+        *list = g_list_append(*list, TARGET_MIME_PNG_ATOM);
+        *list = g_list_append(*list, TARGET_MIME_JPEG_ATOM);
+        *list = g_list_append(*list, TARGET_MIME_TIFF_ATOM);
+        *list = g_list_append(*list, TARGET_MIME_BMP_ATOM);
+    } else if (g_strcmp0(gstring, "application/x-java-file-list") == 0) {
+        *list = g_list_append(*list, TARGET_MIME_URI_LIST_ATOM);
     } else {
-        dnd_source_set_raw(widget, data, atom);
+        *list = g_list_append(*list, gdk_atom_intern(gstring, FALSE));
     }
+    env->ReleaseStringUTFChars(string, gstring);
 }
 
-static void dnd_drag_begin_callback(GtkWidget *widget,
-                                    GdkDragContext *context,
-                                    gpointer user_data)
-{
-    DragView::set_drag_view(widget, context);
+static GList* data_to_targets(JNIEnv *env, jobject data) {
+    jobject keys;
+    jobject keysIterator;
+    jstring next;
+
+    GList *list = NULL;
+
+    init_target_atoms();
+
+    keys = env->CallObjectMethod(data, jMapKeySet, NULL);
+    JNI_EXCEPTION_TO_CPP(env)
+    keysIterator = env->CallObjectMethod(keys, jIterableIterator, NULL);
+    JNI_EXCEPTION_TO_CPP(env)
+    while (env->CallBooleanMethod(keysIterator, jIteratorHasNext) == JNI_TRUE) {
+        next = (jstring)env->CallObjectMethod(keysIterator, jIteratorNext, NULL);
+        JNI_EXCEPTION_TO_CPP(env)
+        add_target_from_jstring(env, &list, next);
+    }
+    return list;
 }
 
-static void dnd_source_push_data(JNIEnv *env, jobject data, jint supported)
-{
+static void dnd_source_push_data(JNIEnv *env, jobject data, jint supported, DragSourceContext *ctx) {
+    GList *targets;
+    if (WindowContextBase:: sm_mouse_drag_window == NULL) {
+        ERROR0("Drag not started on source window.");
+        return;
+    }
```

Comment:
Instead of using a offscreen window to pass events, it uses the window were the drag started, which seems the intended way.